agent/engine/containermetadata/write_metadata.go

```diff
@@ -0,0 +1,168 @@
+	"github.com/cihub/seelog"
+
+	docker "github.com/fsouza/go-dockerclient"
+)
+
+const (
+	inspectContainerTimeout = 30 * time.Second
+	metadataFile            = "metadata.json"
+)
+
+// dockerClient is a wrapper for the docker interface functions we need
+type dockerClient interface {
+	InspectContainer(string, time.Duration) (*docker.Container, error)
+}
+
+// getTaskIDfromArn parses a task Arn and produces the task ID
+func getTaskIDfromArn(taskarn string) string {
+	colonSplitArn := strings.SplitN(taskarn, ":", 6)
+	// Incorrectly formatted Arn (Should not happen)
+	if len(colonSplitArn) < 6 {
+		return ""
+	}
+	arnTaskPartSplit := strings.SplitN(colonSplitArn[5], "/", 2)
+	// Incorrectly formatted Arn (Should not happen)
+	if len(arnTaskPartSplit) < 2 {
+		return ""
+	}
+	return arnTaskPartSplit[1]
+}
+
+// getMetadataFilePath gives the metadata file path for any agent-managed container
+func getMetadataFilePath(task *api.Task, container *api.Container, dataDir string) string {
+	taskID := getTaskIDfromArn(task.Arn)
+	// Empty task ID indicates malformed Arn (Should not happen)
+	if taskID == "" {
+		return ""
+	}
+	return fmt.Sprintf("%s/metadata/%s/%s/", dataDir, taskID, container.Name)
+}
+
+// writeToMetadata writes given data into the metadata file
+func writeToMetadata(task *api.Task, container *api.Container, data []byte, dataDir string) error {
+	mdFilePath := getMetadataFilePath(task, container, dataDir) + metadataFile
+
+	mdfile, err := os.OpenFile(mdFilePath, os.O_WRONLY, 0644)
+	defer mdfile.Close()
+	if err != nil {
+		return err
+	}
+	_, err = mdfile.Write(data)
+	return err
+}
+
+// initMetadataFile initializes metadata file and populates it with initially available
+// metadata about the container's task and instance
+func initMetadataFile(cfg *config.Config, task *api.Task, container *api.Container) (string, error) {
+	// Create task and container directories if they do not yet exist
+	mdDirectoryPath := getMetadataFilePath(task, container, cfg.DataDir)
+	err := os.MkdirAll(mdDirectoryPath, os.ModePerm)
+	if err != nil {
+		return "", err
+	}
+
+	// Create metadata file
+	mdFilePath := mdDirectoryPath + metadataFile
+	err = ioutil.WriteFile(mdFilePath, nil, 0644)
+	if err != nil {
+		return mdFilePath, err
+	}
+
+	// Get common metadata of all containers of this task and write it to file
+	md := acquireStaticMetadata(cfg, task)
+	data, err := json.MarshalIndent(*md, "", "\t")
+	if err != nil {
+		return mdFilePath, err
+	}
+	return mdFilePath, writeToMetadata(task, container, data, cfg.DataDir)
+}
+
+// CreateMetadata creates the metadata file and adds the metadata directory to
+// the container's mounted host volumes
+func CreateMetadata(cfg *config.Config, binds *[]string, task *api.Task, container *api.Container) error {
+	metadataPath, err := initMetadataFile(cfg, task, container)
+	if err != nil {
+		seelog.Errorf("Failed to create metadata file at %s. Error: %s", metadataPath, err.Error())
```

Comment:
Can you switch to errors.Wrap here ?
Have a look [here](https://godoc.org/github.com/pkg/errors)